File: C#/TheOpenMathLibrary.GeneralMathematics/WaveTheory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TheOpenMathLibrary.GeneralMathematics
{
    /// <summary>
    /// Wave theory - the study of the properties of waves
    /// </summary>
    public class WaveTheory
    {
        /// <summary>
        /// Wavelength function 
        /// </summary>
        /// <param name="waveSpeed"></param>
        /// <param name="frequency"></param>
        /// <returns></returns>
        public static double Wavelength(double waveSpeed, double frequency)
        {
            double wavelength = 0;
            wavelength = waveSpeed / frequency;
            return wavelength;
        }

        /// <summary>
        /// Wave vector function 
        /// </summary>
        /// <param name="waveSpeed"></param>
        /// <param name="frequency"></param>
        /// <returns></returns>
        public static double WaveVector(double waveSpeed, double frequency)
        {
            double waveVector = 0;
            waveVector = 2 * 3.14159 * frequency / waveSpeed;
            return waveVector;
        }

        /// <summary>
        /// Frequency function 
        /// </summary>
        /// <param name="waveSpeed"></param>
        /// <param name="wavelength"></param>
        /// <returns></returns>
        public static double Frequency(double waveSpeed, double wavelength)
        {
            double frequency = 0;
            frequency = waveSpeed / wavelength;
            return frequency;
        }

        /// <summary>
        /// Angular frequency function 
        /// </summary>
        /// <param name="frequency"></param>
        /// <returns></returns>
        public static double AngularFrequency(double frequency)
        {
            double angularFrequency = 0;
            angularFrequency = 2 * 3.14159 * frequency;
            return angularFrequency;
        }

        /// <summary>
        /// Oscillatory velocity 
        /// </summary>
        /// <param name="amplitude"></param>
        /// <param name="angularFrequency"></param>
        /// <param name="time"></param>
        /// <param name="phase"></param>
        /// <returns></returns>
        public static double OscillatoryVelocity(double amplitude, double angularFrequency, double time, double phase)
        {
            double oscillatoryVelocity = 0;
            oscillatoryVelocity = amplitude * Math.Sin(angularFrequency * time + phase);
            return oscillatoryVelocity;
        }

        /// <summary>
        /// Oscillatory acceleration 
        /// </summary>
        /// <param name="amplitude"></param>
        /// <param name="angularFrequency"></param>
        /// <param name="time"></param>
        /// <param name="phase"></param>
        /// <returns></returns>
        public static double OscillatoryAcceleration(double amplitude, double angularFrequency, double time, double phase)
        {
            double oscillatoryAcceleration = 0;
            oscillatoryAcceleration = amplitude * angularFrequency * Math.Cos(angularFrequency * time + phase);
            return oscillatoryAcceleration;
        }

        /// <summary>
        /// Phase velocity 
        /// </summary>
        /// <param name="waveSpeed"></param>
        /// <param name="waveVector"></param>
        /// <returns></returns>
        public static double PhaseVelocity(double waveSpeed, double waveVector)
        {
            double phaseVelocity = 0;
            phaseVelocity = waveSpeed / waveVector;
            return phaseVelocity;
        }

        /// <summary>
        /// Group velocity 
        /// </summary>
        /// <param name="waveSpeed"></param>
        /// <param name="waveVector"></param>
        /// <returns></returns>
        public static double GroupVelocity(double waveSpeed, double waveVector)
        {
            double groupVelocity = 0;
            groupVelocity = waveSpeed / waveVector;
            return groupVelocity;
        }

        /// <summary>
        /// Time delay 
        /// </summary>
        /// <param name="time"></param>
        /// <param name="phase"></param>
        /// <returns></returns>
        public static double TimeDelay(double time, double phase)
        {
            double timeDelay = 0;
            timeDelay = time - phase;
            return timeDelay;
        }

        /// <summary>
        /// Phase difference 
        /// </summary>
        /// <param name="phase1"></param>
        /// <param name="phase2"></param>
        /// <returns></returns>
        public static double PhaseDifference(double phase1, double phase2)
        {
            double phaseDifference = 0;
            phaseDifference = phase1 - phase2;
            return phaseDifference;
        }

        /// <summary>
        /// Phase function
        /// </summary>
        /// <param name="waveVector"></param>
        /// <param name="position"></param>
        /// <param name="time"></param>
        /// <param name="phase"></param>
        /// <returns></returns>
        public static double Phase(double waveVector, double position, double time, double phase)
        {
            phase = waveVector * position - 2 * 3.14159 * time + phase;
            return phase;
        }
    }
}

File: Go/TheOpenMathLibrary/GeneralMathematics/WaveTheory.go
package generalmathematics

import (
	"math"
)

// WaveTheory provides various wave theory functions.
type WaveTheory struct{}

// Wavelength returns the wavelength given wave speed and frequency.
func (WaveTheory) Wavelength(waveSpeed, frequency float64) float64 {
	return waveSpeed / frequency
}

// WaveVector returns the wave vector given wave speed and frequency.
func (WaveTheory) WaveVector(waveSpeed, frequency float64) float64 {
	return 2 * math.Pi * frequency / waveSpeed
}

// Frequency returns the frequency given wave speed and wavelength.
func (WaveTheory) Frequency(waveSpeed, wavelength float64) float64 {
	return waveSpeed / wavelength
}

// AngularFrequency returns the angular frequency given frequency.
func (WaveTheory) AngularFrequency(frequency float64) float64 {
	return 2 * math.Pi * frequency
}

// OscillatoryVelocity returns the oscillatory velocity.
func (WaveTheory) OscillatoryVelocity(amplitude, angularFrequency, time, phase float64) float64 {
	return amplitude * math.Sin(angularFrequency*time+phase)
}

// OscillatoryAcceleration returns the oscillatory acceleration.
func (WaveTheory) OscillatoryAcceleration(amplitude, angularFrequency, time, phase float64) float64 {
	return amplitude * angularFrequency * math.Cos(angularFrequency*time+phase)
}

// PhaseVelocity returns the phase velocity given wave speed and wave vector.
func (WaveTheory) PhaseVelocity(waveSpeed, waveVector float64) float64 {
	return waveSpeed / waveVector
}

// GroupVelocity returns the group velocity given wave speed and wave vector.
func (WaveTheory) GroupVelocity(waveSpeed, waveVector float64) float64 {
	return waveSpeed / waveVector
}

// TimeDelay returns the time delay given time and phase.
func (WaveTheory) TimeDelay(time, phase float64) float64 {
	return time - phase
}

// PhaseDifference returns the phase difference given two phases.
func (WaveTheory) PhaseDifference(phase1, phase2 float64) float64 {
	return phase1 - phase2
}

// Phase returns the phase given wave vector, position, time, and initial phase.
func (WaveTheory) Phase(waveVector, position, time, phase float64) float64 {
	return waveVector*position - 2*math.Pi*time + phase
}
